### DISCUSSION TITLE
Nuxt server init implementation

### DISCUSSION THREAD

**Discussion by silencerspirit** [nuxt/nuxt#15886] [Questions]
Hi there!

which is better way to implement nuxtServerInit in nuxtjs3?
1) **/plugins/server-init.server.ts**
`export default defineNuxtPlugin(() => {
  console.log('CLIENT-INIT-PLUGIN');
});
`

2) **middleware/server-init.ts**
`
export default function () {
    if (process.server) {
        console.log('CLIENT-INIT-PLUGIN');
    }
}
`
**nuxt.config.ts**
`{ router: { middleware: ['nuxt-server-init'] } }`

P.S. I'm still leaning towards option 1, but here I get warning:
_[Vue warn]: onServerPrefetch is called when there is no active component instance to be associated with. Lifecycle injection APIs can only be used during execution of setup(). If you are using async setup(), make sure to register lifecycle hooks before the first await statement._

**Comment by danielroe** [comment]
What are you running within the code? I'm guessing it's probably a call to pinia, which means you should probably run this within your `app.vue` setup function instead, as pinia currently requires to be called within the component lifecycle.

**Reply by silencerspirit** [reply]
@danielroe i'm calling `useFetch()` method, and `$patch` store method.
so code looks like this: 
**/plugins/server-init.server.ts**
```
import { useIndexStore } from '@/store'; 
 export default defineNuxtPlugin(async () => {
  const store = useIndexStore();
  const { data } = await useFetch("...url");
  store.$patch((state) => (state.someData = data));
});
```

It's works correctly but i'm not sure this is right way implementation 'nuxtServerInit by vuex')

**Comment by noopurphalak** [comment]
@silencerspirit I understood how you are just calling this plugin on the server. If you need to refresh the data on the consequent renders, would you again call `useFetch` in a store action to fetch the updated posts? I am a little bit confused about that part...

PS: @danielroe You can give your thought on the same.

**Reply by silencerspirit** [reply]
Hello! I don't quite understand what you mean. If you want to re-request your useFetch, then you can call the **refresh()** method which returns useFetch.

As for my final implementation of nuxtServerInit, I just settled on using **useAsyncData** inside my **app.vue** as advised @danielroe

**Comment by tol64** [comment]
@danielroe, @silencerspirit

When a web application with server rendering (**SSR** + **SPA**) has many pages and layouts, `nuxtServerInit()` was a good option to load and prepare data for the whole application in one place and make sure to do it at the very first application load or page refresh. That is, before all the plugins and middlewares are executed.

So the `plugins/server-init.server.ts` option looks like a good replacement:
```
import { useMainStore } from "../stores/index";

export default defineNuxtPlugin(async () => {
  if (!process.server)
    return;

  const store = useMainStore();
  // ...
});
```

But then it must be the only plugin to run on the server side (`plugins/*.server.ts`)? Otherwise, if there will be several of them, how to specify which one should be executed first? Is it possible to specify priority for execution of plugins?

**Reply by miker059** [reply]
Plugins are called in alphabetical order, in order to set the call sequence, just put a number in front of the plugin name, for example, 1.myPlugin.ts and so on.

**Comment by tol64** [comment]
I also decided to apply the following scheme in the `app.vue` file for now. The first tests show that this is what we need. 
```
<template>
  <NuxtLayout>
    <NuxtPage/>
  </NuxtLayout>
</template>

<script setup lang="ts">
import { useMainStore } from "~/stores/index";

const store = useMainStore();

onServerPrefetch(async () => {
  // ...
})
</script>
```

**Reply by ilyaDegtyarenko** [reply]
This option would be nice, but the script will be executed after plugins and middleware.

For example, user data is required in a plugin, especially in middleware

**Comment by ndragun92** [comment]
This is how I handle this. I create a plugin that I call a composable which is saving the value in the store once I receive data

`/plugins/001_init.server.ts`
```typescript
import useApiOrg from "~/composables/api/org/useApiOrg";

export default defineNuxtPlugin(async (nuxtApp) => {
  console.log("001_init.server.ts");
  const { get } = useApiOrg();
  await get();
  const { token } = useAuth();
  if (token.value) {
    const { getProfile } = useUser();
    await getProfile();
  }
});

```

`/composables/api/org/useApiOrg.ts`

```typescript
import { useOrgStore } from "~/store/orgStore";
import type { ApiOrgInterface } from "~/composables/api/org/ApiOrg.interface";
import useApiBaseConfig from "~/composables/api/useApiBaseConfig";

export default function useApiOrg() {
  const orgStore = useOrgStore();
  const { serverHeaders } = useApiBaseConfig();

  const get = async () => {
    try {
      const data: ApiOrgInterface[] = await $fetch("/api/orgs", {
        headers: {
          ...serverHeaders(),
        },
      });
      if (data?.length) {
        orgStore.setSettings(data[0] as ApiOrgInterface);
      }
    } catch (e) {
      throw createError({
        statusCode: 404,
        statusMessage: "Missing org settings",
      });
    }
  };

  return { get };
}
```

**Comment by cryptovibe1** [comment]
`server/plugins/001.serverInit.ts`

```
export default defineNitroPlugin(async () => {
  console.log('Init Server')
})
```
